### PR TITLE
BE-011 Hanna Kalashnik once coupon/fave create/del by discount Id feature

### PR DESCRIPTION
**Pull request by Hanna-Kalashnyk**
Coupon/favorite post/del once by discountId
- redundant methods

### CODE REVIEW

File: src/main/java/com/exadel/discount/controller/CouponController.java [hunk 1]
```diff
@@ -89,7 +87,7 @@ public CouponDTO getCouponById(@PathVariable @NotNull final UUID id) {
     @PostMapping
     @ApiOperation("Save new coupon to user")
     @UserAccess
-    public CouponDTO addCoupon(@RequestBody @NotNull final CreateCouponDTO createCouponDTO) {
```
Comment (el-dub) [inline src/main/java/com/exadel/discount/controller/CouponController.java:92]: If you don't need CreateCouponDTO and CreateFavoriteDTO anymore, you can delete these files

File: src/main/java/com/exadel/discount/repository/CouponRepository.java [hunk 2]
```diff
@@ -37,4 +33,8 @@
     @EntityGraph(attributePaths = {"discount", "discount.category", "discount.vendorLocations", "discount.tags",
             "discount.vendor", "discount.vendorLocations.city", "discount.vendorLocations.city.country"})
     List<Coupon> findAllByIdIn(Iterable<UUID> ids, Sort sort);
+
+    @EntityGraph(attributePaths = {"user", "discount"})
+    Optional<Coupon> findByDiscountIdAndAndUserEmail(@Param("discountId") UUID discountId,
+                                                     @Param("userEmail") String userEmail);
```
Comment (el-dub) [inline src/main/java/com/exadel/discount/repository/CouponRepository.java]: You don't need @Param if you don't use @Query

File: src/main/java/com/exadel/discount/repository/FavoriteRepository.java [hunk 3]
```diff
@@ -32,4 +34,17 @@
     @EntityGraph(attributePaths = {"discount", "discount.category", "discount.vendorLocations", "discount.tags",
             "discount.vendor", "discount.vendorLocations.city", "discount.vendorLocations.city.country"})
     List<Favorite> findAllByIdIn(Iterable<UUID> ids, Sort sort);
+
+    @Modifying
+    void deleteFavoriteByDiscountIdAndUserEmail(@Param("discountId") UUID discountId,
+                                                @Param("userEmail") String userEmail);
```
Comment (el-dub) [inline src/main/java/com/exadel/discount/repository/FavoriteRepository.java]: Remove @Param please

File: src/main/java/com/exadel/discount/service/impl/FavoriteServiceImpl.java [hunk 4]
```diff
@@ -83,39 +85,46 @@ public FavoriteDTO findFavoriteById(UUID id) {
...
     @Override
-    public void deleteFavoriteByID(UUID id) {
-        log.debug("Finding & deleting Favorite by ID");
+    public void deleteFavoriteByID(UUID discountId) {
+        log.debug("Finding & deleting Favorite by DiscountID");
+        String userEmail = SecurityContextHolder.getContext().getAuthentication().getName();
+        if (favoriteRepository.findByDiscountIdAndAndUserEmail(discountId, userEmail).isEmpty()) {
+            throw new DeletionRestrictedException(String
+                    .format("Favorite for discount with id %s does not already exist", discountId));
+        }
```
Comment (el-dub) [inline src/main/java/com/exadel/discount/service/impl/FavoriteServiceImpl.java]: You can use existsByDiscountIdAndUserEmail, it'll return boolean

File: src/main/java/com/exadel/discount/service/impl/FavoriteServiceImpl.java [hunk 5]
```diff
@@ -83,39 +85,46 @@ public FavoriteDTO findFavoriteById(UUID id) {
...
     @Override
-    public FavoriteDTO assignFavoriteToUser(CreateFavoriteDTO createFavoriteDTO) {
-        log.debug("Finding of certain user by ID");
+    public FavoriteDTO assignFavoriteToUser(UUID discountId) {
+        log.debug(String.format("Check: if user already have Favorite of Discount with ID %s ", discountId));
+        String userEmail = SecurityContextHolder.getContext().getAuthentication().getName();
+        if (favoriteRepository.findByDiscountIdAndAndUserEmail(discountId, userEmail).isPresent()) {
+            throw  new CreationRestrictedException(String
+                    .format("Favorite for discount with id %s does already exist", discountId));
+        }
```
Comment (el-dub) [inline src/main/java/com/exadel/discount/service/impl/FavoriteServiceImpl.java]: You can use existsByDiscountIdAndUserEmail, it'll return boolean

File: src/main/java/com/exadel/discount/service/impl/CouponServiceImpl.java [hunk 6]
```diff
@@ -83,28 +84,31 @@ public CouponDTO findCouponById(UUID id) {
...
-    public CouponDTO assignCouponToUser(CreateCouponDTO createCouponDTO) {
-        log.debug("Finding of certain user by ID");
-
+    public CouponDTO assignCouponToUser(UUID discountId) {
+        log.debug(String.format("Check: if user already have Favorite of Discount with ID %s ", discountId));
+        String userEmail = SecurityContextHolder.getContext().getAuthentication().getName();
+        if (couponRepository.findByDiscountIdAndAndUserEmail(discountId, userEmail).isPresent()) {
+            throw new CreationRestrictedException(String
+                    .format("Coupon for Discount with id %s does already exist", discountId));
+        }
```
Comment (el-dub) [inline src/main/java/com/exadel/discount/service/impl/CouponServiceImpl.java]: You can use existsByDiscountIdAndUserEmail, it'll return boolean

File: src/main/java/com/exadel/discount/repository/FavoriteRepository.java [hunk 7]
```diff
@@ -32,4 +34,17 @@
...
+
+
+    @EntityGraph(attributePaths = {"user", "discount"})
+    Optional<Favorite> findByDiscountIdAndAndUserEmail(@Param("discountId") UUID discountId,
+                                                       @Param("userEmail") String userEmail);
+
+    @EntityGraph(attributePaths = {"user", "discount"})
+    boolean existsFavoriteByDiscountIdAndAndUserEmail(@Param("discountId") UUID discountId,
+                                                      @Param("userEmail") String userEmail);
```
Comment (el-dub) [inline src/main/java/com/exadel/discount/repository/FavoriteRepository.java]: Just existsByDiscountIdAndUserEmail, you have typo in naming and you don't need EntityGraph here

File: src/main/java/com/exadel/discount/service/impl/FavoriteServiceImpl.java [hunk 8]
```diff
@@ -83,39 +85,46 @@ public FavoriteDTO findFavoriteById(UUID id) {
...
+        log.debug(String.format("Successfully added Favorite for Discount with ID %s to user", discountId));
 
         return favoriteMapper.toFavoriteDTO(favoriteSaved);
     }
 
     @Transactional
     @Override
-    public void deleteFavoriteByID(UUID id) {
-        log.debug("Finding & deleting Favorite by ID");
+    public void deleteFavoriteByID(UUID discountId) {
```
Comment (el-dub) [inline src/main/java/com/exadel/discount/service/impl/FavoriteServiceImpl.java]:
```suggestion
    public void deleteFavoriteByDiscountID(UUID discountId) {
```

**Review by el-dub — CHANGES_REQUESTED** [review]

File: src/main/java/com/exadel/discount/controller/CouponController.java (hunk 1 above)
Comment (Hanna-Kalashnyk) [reply src/main/java/com/exadel/discount/controller/CouponController.java:92]: Done

File: src/main/java/com/exadel/discount/repository/CouponRepository.java (hunk 2 above)
Comment (Hanna-Kalashnyk) [reply src/main/java/com/exadel/discount/repository/CouponRepository.java]: Done

File: src/main/java/com/exadel/discount/repository/FavoriteRepository.java (hunk 3 above)
Comment (Hanna-Kalashnyk) [reply src/main/java/com/exadel/discount/repository/FavoriteRepository.java]: Done

File: src/main/java/com/exadel/discount/service/impl/FavoriteServiceImpl.java (hunk 4 above)
Comment (Hanna-Kalashnyk) [reply src/main/java/com/exadel/discount/service/impl/FavoriteServiceImpl.java]: Done

File: src/main/java/com/exadel/discount/service/impl/FavoriteServiceImpl.java (hunk 5 above)
Comment (Hanna-Kalashnyk) [reply src/main/java/com/exadel/discount/service/impl/FavoriteServiceImpl.java]: Done

File: src/main/java/com/exadel/discount/service/impl/CouponServiceImpl.java (hunk 6 above)
Comment (Hanna-Kalashnyk) [reply src/main/java/com/exadel/discount/service/impl/CouponServiceImpl.java]: Done

File: src/main/java/com/exadel/discount/service/impl/FavoriteServiceImpl.java (hunk 4 above)
Comment (el-dub) [reply src/main/java/com/exadel/discount/service/impl/FavoriteServiceImpl.java]: I've checked - it seems that everything works. Just use correct naming and remove Entity Graph

File: src/main/java/com/exadel/discount/repository/FavoriteRepository.java (hunk 7 above)
Comment (Hanna-Kalashnyk) [reply src/main/java/com/exadel/discount/repository/FavoriteRepository.java]: Done

File: src/main/java/com/exadel/discount/service/impl/FavoriteServiceImpl.java (hunk 8 above)
Comment (Hanna-Kalashnyk) [reply src/main/java/com/exadel/discount/service/impl/FavoriteServiceImpl.java]: Done

**Review by el-dub — APPROVED** [review]

**Review by SergeyKoval — APPROVED** [review]

**Pull request merged**